cmd/stdiscosrv/database.go

```diff
@@ -0,0 +1,264 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//go:generate go run ../../script/protofmt.go database.proto
+//go:generate protoc -I ../../../../../ -I ../../vendor/ -I ../../vendor/github.com/gogo/protobuf/protobuf -I . --gogofast_out=. database.proto
+
+package main
+
+import (
+	"errors"
+	"sort"
+	"time"
+
+	"github.com/syndtr/goleveldb/leveldb/util"
+
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+type clock interface {
+	Now() time.Time
+}
+
+type defaultClock struct{}
+
+func (defaultClock) Now() time.Time {
+	return time.Now()
+}
+
+type database interface {
+	put(key string, rec DatabaseRecord) error
+	get(key string) (DatabaseRecord, error)
+}
+
+var errExpired = errors.New("expired")
+
+type levelDBStore struct {
+	db         *leveldb.DB
+	inbox      chan func()
+	stop       chan struct{}
+	clock      clock
+	marshalBuf []byte
+}
+
+func newLevelDBStore(dir string) (*levelDBStore, error) {
+	db, err := leveldb.OpenFile(dir, levelDBOptions)
+	if err != nil {
+		return nil, err
+	}
+	return &levelDBStore{
+		db:    db,
+		inbox: make(chan func(), 16),
+		stop:  make(chan struct{}),
+		clock: defaultClock{},
+	}, nil
+}
+
+func (s *levelDBStore) put(key string, rec DatabaseRecord) error {
+	t0 := time.Now()
+	rc := make(chan error)
+	s.inbox <- func() {
+		if len(rec.Addresses) != 0 {
+			rec.Misses = 0
+			// grab the existing record
+			cur, err := s.get(key)
+			if err == nil {
+				// merge addresses if the get was successful
+				databaseMerges.Inc()
+				rec.Addresses = merge(cur.Addresses, rec.Addresses)
+			}
+		}
+		size := rec.Size()
+		if len(s.marshalBuf) < size {
```

Comment:
It's done that way because the slice is never resliced, as MarshalTo just overwrites the first part of it, cap is always == len and len is thus less magical